lib/stitch_fix/log_weasel/transaction.rb

```diff
@@ -1,15 +1,11 @@
-begin
-  require 'securerandom'
-rescue
-  require 'active_support/secure_random'
-end
+require "ulid"
 
 module StitchFix
   module LogWeasel
     module Transaction
 
       def self.create(key = nil)
-        Thread.current[:log_weasel_id] = "#{key ? "#{key}-" : ""}#{SecureRandom.hex(10)}"
+        Thread.current[:log_weasel_id] = "#{ULID.generate}#{key ? "-#{key}" : ""}"
```

Comment:
nice

Reply:
Actually I just thought of a question (more of a realization) about this change. Is it true that the value in having a correlation ID that is sortable by time is really that it allows you to order logs for _different_ units of work? Using ULIDs here will still result in having a static, unchanging, correlation ID for a _single_ unit of work across services, but the ability to sort the correlation IDs will help us understand the order in which operations happen, which our current logs lack (if you don't believe the timestamps). But this is really orthogonal to being able to trace a unit of work across application boundaries in the logs. Am I missing something?

Reply:
Current reality for reference: https://production-stitchfix.datadoghq.com/logs?cols=%5B%5D&event&from_ts=1549948868234&index=main&live=true&query=%40log.log_weasel_trace_id%3AHELLBLAZER-WEB-48e1432b1a7772009133+OR+HELLBLAZER-WEB-48e1432b1a7772009133&stream_sort=desc&to_ts=1550121668234

This log shows the log weasel ID across service boundaries in order, from newest at the top to oldest at the bottom. A ULID doesn't really add anything in this view, as the time element would be static for this unit of work.

Reply:
I don't think you are missing anything. There is certainly no need to sort the same ID when correlating a single request across the system. However if there is a charge (originating in an iphone checkout) and a refund (originating in Kingmob) it would be ideal to be to see which request occurred first.  Another case would be to to sort signup requests (such as in our recent attacks). Another would be to understand how long it took for a request takes from initiation to processing at particular points across a user initiated _transaction_.

These are pretty contrived examples, sorry. But I think they convey intent. So it is arguable what the best sort order is for different situations. I like the idea that the sorting is ordered by time of _transaction initiation_ for the cases where we ever want to compare things represented by the IDs themselves. I think that cannot happen if the are prefixed, at least is would be less simple with little gain. For the most part the IDs are universally unique and opaque, however if we ever need to ask questions about them in a structured way (an index or db column, e.g.) being able to order them meaningfully -- as compared to themselves -- would prove significant. 

So that was a bunch of words. Let's try this...
> Is it true that the value in having a correlation ID that is sortable by time is really that it allows you to order logs for _different_ units of work?
Yes
> But this is really orthogonal to being able to trace a unit of work across application boundaries in the logs
Agreed.
> Am I missing something?
I don't think so.

Reply:
Perfect, thanks for your answer. I think I was trying to answer the question in my own head of what problem do ULIDs solve here, and your examples make a lot of sense. I'm not sure this aspect was spelled out in the doc you wrote up: https://docs.google.com/document/d/1wPn7M2HE2ghZJutWaMbctNcSroc-9r1FFdEUriYlxDc/edit

but it might be a nice addition.